README.markdown
[![Build Status](https://travis-ci.org/seanh/flitter.svg)](https://travis-ci.org/seanh/flitter)
[![Coverage Status](https://img.shields.io/coveralls/seanh/flitter.svg)](https://coveralls.io/r/seanh/flitter)
[![Latest Version](https://pypip.in/version/flitter/badge.svg)](https://pypi.python.org/pypi/flitter/)
[![Downloads](https://pypip.in/download/flitter/badge.svg)](https://pypi.python.org/pypi/flitter/)
[![Supported Python versions](https://pypip.in/py_versions/flitter/badge.svg)](https://pypi.python.org/pypi/flitter/)
[![Development Status](https://pypip.in/status/flitter/badge.svg)](https://pypi.python.org/pypi/flitter/)
[![License](https://pypip.in/license/flitter/badge.svg)](https://pypi.python.org/pypi/flitter/)


Flitter
=======

Flitter makes launching apps and switching between windows as fast and easy
as possible:

* F1 launches Firefox, or focuses the Firefox window if Firefox is already
  running.
* If there's more than one Firefox window open, repeatedly hitting F1 cycles
  through them.
* F2 does the same for gVim.
* And so on, binding all your most-frequently used apps to the function keys
  or other keyboard shortcuts (you configure the apps and shortcuts yourself).
* I have F1..F10 bound to my 10 most frequently used apps, and F11 cycling
  through all other windows that don't belong to my top ten apps.
* When moving between apps Flitter raises each app's most recently used window
  first, and when cycling through an app's windows it goes through them in
  most-recently-used-first order.

Compared to using the mouse or traditional fast window switching shortcuts
(like Alt-Tab), with Flitter:

* You don't need to keep track of which apps are open and closed.
  Let the computer do that. F1 just takes you to Firefox, opening it if
  necessary. You can only Alt-Tab to an app if it's open, if not you have to
  do something else to launch it.
* You never need to use the mouse to open apps or switch windows.
* You never need to use two hands or two finger contortions to press multiple
  keys at once (Alt+Tab, Tab, Tab...). Just hit one function key.
* You only need to use one key (perhaps hitting it
  repeatedly) to get to a window, no Alt+Tab,Tab,Tab to get to an app then
  Alt+\`,\`,\` or Alt+down,left,left to get to the window.
* You don't need to look at or think about anything on screen other than the
  app windows themselves as they're focused (no finding the right icon in an
  Alt-Tab dialog)
* Very finger memory compatible, your hands will quickly memorize F1 for
  Firefox, F5 for Thunderbird, F8 for Skype, and you'll be switching apps at
  the speed of thought.

Repeatedly hitting F1 to cycle through Firefox windows doesn't scale well if
you have dozens of Firefox windows open.
But personally I usually have just one, and never more than three or four,
windows per app (and then sometimes several tabs within each window) and
Flitter works great for me (focusing the most recently used windows first makes
a big difference).

Binding each function key to an app doesn't scale when you have more apps than
function keys. You can just fall back on Alt-Tab for apps outside of your top
12, but `flitter --others` (see below) gives you a key for cycling through
windows that don't belong to any of your bound apps. I find this lets me avoid
Alt-Tab entirely.


Requirements
------------

Flitter requires Python 2.7, [wmctrl](http://tomas.styblo.name/wmctrl/) and
works with any WMH/NetWM compatible X Window Manager (Gnome, Unity, Openbox...)

It doesn't work on Windows, OS X, or non-WMH/NetWM linux environments yet,
although porting should be possible (just replace
[wmctrl.py](flitter/wmctrl.py)
with something capable of interacting with your desktop's windows).


Installation
------------

First install wmctrl. On Debian or Ubuntu, just:

    $ sudo apt-get install wmctrl

Then install Flitter:

    $ pip install flitter

You should now be able to run the `flitter` command in your shell.
Run `flitter -h` for help.


Configuration & Usage
---------------------

Copy the [default configuration file](flitter/flitter.json)
to `~/.flitter.json`. This is a [JSON](http://json.org/) file containing a list
of _window specs_. Window specs are how Flitter knows which windows belong to
which app. Each spec has a name, such as `Firefox`, and a number of properties
that are matched against the properties of your open windows to decide whether
each window is a Firefox window or not. For example:

        "Firefox": {
            "wm_class": ".Firefox",
            "command": "firefox"
        },

This window spec will match all windows whose WM_CLASS property contains the
string ".Firefox" (in other words, all Firefox windows).

To have Flitter raise a Firefox window or launch Firefox, run it with the
spec's name as the command-line argument:

    $ flitter firefox

Flitter doesn't have built-in support for keyboard shortcuts.
You just use whatever mechanism your window manager provides to bind keyboard
shortcuts to flitter commands.

To see a list of all your open windows and their properties so you can write
window specs for them, run `wmctrl -lxp` (see `man wmctrl` for more info).

The `"command"` part of the spec is the command that Flitter will run to launch
Firefox, if it finds no Firefox windows.

This way of identifying windows is quite flexible. You can go beyond the simple
one app per keyboard shortcut model, for example:

* A key to switch to the Firefox window showing Google Calender or open Google
  Calender in a new Firefox window
* A key to switch to the Gnome Terminal window running WeeChat or open a new
  Gnome Terminal window with the WeeChat profile

The full set of attributes that you can include in a window spec is:

`window_id`
  The unique ID of an open window, e.g. 0x0180000b

`desktop`
  The desktop that the window is on, e.g. 1

`pid`
  The process ID of the window, e.g. 3384

`wm_class`
  The WM_CLASS of the window, e.g. Navigator.Firefox

`machine`
  The client machine that the window belongs to

`title`
  The window title


Development Install
-------------------

To install Flitter in a virtual environment for development first install
virtualenv, on Debian or Ubuntu do:

    $ sudo apt-get install python-virtualenv

Then create and activate a Python virtual environment and install Flitter into
it:

    $ virtualenv flitter
    $ . flitter/bin/activate
    $ cd flitter
    $ pip install -e 'git+https://github.com/seanh/flitter.git#egg=flitter'

`which flitter` should now report the flitter binary in your virtualenv.

To run the tests do:

    $ cd src/flitter
    $ pip install -r dev-requirements.txt
    $ nosetests

To run the tests and produce a test coverage report, do:

    $ nosetests --with-coverage --cover-inclusive --cover-erase --cover-tests

To upload a new release of Flitter to PyPI ans GitHub:

1. Update the version number in [setup.py](setup.py).
2. `python setup.py sdist`
3. `python setup.py sdist upload`
4. `git commit setup.py -m "Release version X.Y.Z"`
5. `git tag X.Y.Z`
6. `git push`
7. `git push --tags`


CHANGELOG.txt
v1.0.1:
- Use the ewhm Python library instead of wmctrl, so there are now no
  non-Python dependencies
- Add --current-desktop option
- Add --ignore-minimized option
- Use regexes to match window specs
- Add --print-matching option
- Fix some crashes


test_wmctrl.py
"""Tests for wmctrl.py."""

import mock

import wmctrl


def test_parse_wmctrl_window_line():
    """Test that _parse_wmctrl_window_line() returns the right result for a
    typical wmctrl -lxp output line.

    """
    expected_result = dict(
        window_id='0x042000b3',
        desktop='0',
        pid='4904',
        wm_class='Mail.Thunderbird',
        machine='mistakenot',
        title='Inbox - Unified Folders - Mozilla Thunderbird')
    assert wmctrl._parse_wmctrl_window_line(
        '0x042000b3  0 4904   Mail.Thunderbird      mistakenot Inbox - Unified '
        'Folders - Mozilla Thunderbird') == expected_result


def test_parse_wmctrl_window_line_with_spaces_in_wm_class():
    """Test that _parse_wmctrl_window_line() returns the right result when the
    wmctrl -lxp output line has a wm_class with spaces in it.

    The wmctrl -lxp output line items are separated by spaces, but the wm_class
    item may contain spaces so it's separated by double spaces on either side.

    """
    expected_result = dict(
        window_id='0x04400051',
        desktop='0',
        pid='5116',
        wm_class='Time Tracker.Hamster',
        machine='mistakenot',
        title='Time Tracker')
    assert wmctrl._parse_wmctrl_window_line(
        '0x04400051  0 5116   Time Tracker.Hamster  mistakenot '
        'Time Tracker') == expected_result


def test_window_eq_when_equal():
    """__eq__() should return True when given an equal other Window."""
    window_1 = wmctrl.Window('window_id', 'desktop', 'pid', 'wm_class',
                             'machine', 'title')
    window_2 = wmctrl.Window('window_id', 'desktop', 'pid', 'wm_class',
                             'machine', 'title')

    assert window_1 == window_2


def test_window_eq_when_not_equal():
    """__eq__() should return False when given an unequal other Window."""
    window_1 = wmctrl.Window('window_id', 'desktop', 'pid', 'wm_class',
                             'machine', 'title')
    window_2 = wmctrl.Window('window_id_2', 'desktop', 'pid_2', 'wm_class',
                             'machine', 'title')

    assert window_1 != window_2


def test_window_eq_when_not_a_window_object():
    """__eq__() should return False when given a non-Window object."""
    window = wmctrl.Window('window_id', 'desktop', 'pid', 'wm_class',
                           'machine', 'title')
    assert window != object()


def test_windows():
    """Test that _windows() returns the right windows.

    This includes windows with spaces in their WM_CLASS, with non-ASCII
    characters in their window title, etc. so that the wmctrl output parsing
    is tested.

    """
    window_list = open('example_wmctrl_window_list.txt', 'r').read()
    windows = wmctrl._windows(window_list)
    window_ids = [window.window_id for window in windows]
    expected_window_ids = [
        '0x02a00001', '0x03200002', '0x0280000a', '0x03200007', '0x03200011',
        '0x03200016', '0x0320001b', '0x0320001c', '0x02c000b3', '0x03e0000c',
        '0x04000155', '0x042000b3', '0x04400051', '0x04c00001', '0x05000012',
        '0x03e005d4', '0x0580002b', '0x05600003']

    assert window_ids == expected_window_ids


def test_focus():
    """Test that Window._focus() calls _run() as we expect it to."""
    window = wmctrl.Window('test_window_id', 'desktop', 'pid',
                           'Navigator.Firefox', 'machine', 'title')
    mock_run_function = mock.MagicMock()

    window._focus(mock_run_function)

    assert mock_run_function.called_once_with('wmctrl -i -a test_window_id')


def test_focused_window():
    """Test that _focused_window() returns the right Window."""
    # Example output from wmctrl -a :ACTIVE: -v
    wmctrl_output = '''envir_utf8: 1
Using window: 0x03e00008
'''
    matching_window = wmctrl.Window('0x03e00008', 'desktop', 'pid', 'wm_class',
                                    'machine')
    windows = [
        wmctrl.Window('window_id', 'desktop', 'pid', 'wm_class', 'machine'),
        matching_window,
        wmctrl.Window('second_window_id', 'desktop', 'pid', 'wm_class',
                      'machine'),
    ]
    assert wmctrl._focused_window(wmctrl_output, windows) == matching_window


flitter/test_wmctrl.py
"""Tests for wmctrl.py."""

import mock

import flitter.wmctrl as wmctrl


def test_parse_wmctrl_window_line():
    """Test that _parse_wmctrl_window_line() returns the right result for a
    typical wmctrl -lxp output line.

    """
    expected_result = dict(
        window_id='0x042000b3',
        desktop='0',
        pid='4904',
        wm_class='Mail.Thunderbird',
        machine='mistakenot',
        title='Inbox - Unified Folders - Mozilla Thunderbird')
    assert wmctrl._parse_wmctrl_window_line(
        '0x042000b3  0 4904   Mail.Thunderbird      mistakenot Inbox - Unified '
        'Folders - Mozilla Thunderbird') == expected_result


def test_parse_wmctrl_window_line_with_spaces_in_wm_class():
    """Test that _parse_wmctrl_window_line() returns the right result when the
    wmctrl -lxp output line has a wm_class with spaces in it.

    The wmctrl -lxp output line items are separated by spaces, but the wm_class
    item may contain spaces so it's separated by double spaces on either side.

    """
    expected_result = dict(
        window_id='0x04400051',
        desktop='0',
        pid='5116',
        wm_class='Time Tracker.Hamster',
        machine='mistakenot',
        title='Time Tracker')
    assert wmctrl._parse_wmctrl_window_line(
        '0x04400051  0 5116   Time Tracker.Hamster  mistakenot '
        'Time Tracker') == expected_result


def test_window_eq_when_equal():
    """__eq__() should return True when given an equal other Window."""
    window_1 = wmctrl.Window('window_id', 'desktop', 'pid', 'wm_class',
                             'machine', 'title')
    window_2 = wmctrl.Window('window_id', 'desktop', 'pid', 'wm_class',
                             'machine', 'title')

    assert window_1 == window_2


def test_window_eq_when_not_equal():
    """__eq__() should return False when given an unequal other Window."""
    window_1 = wmctrl.Window('window_id', 'desktop', 'pid', 'wm_class',
                             'machine', 'title')
    window_2 = wmctrl.Window('window_id_2', 'desktop', 'pid_2', 'wm_class',
                             'machine', 'title')

    assert window_1 != window_2


def test_window_eq_when_not_a_window_object():
    """__eq__() should return False when given a non-Window object."""
    window = wmctrl.Window('window_id', 'desktop', 'pid', 'wm_class',
                           'machine', 'title')
    assert window != object()


def test_windows():
    """Test that _windows() returns the right windows.

    This includes windows with spaces in their WM_CLASS, with non-ASCII
    characters in their window title, etc. so that the wmctrl output parsing
    is tested.

    """
    window_list = open('example_wmctrl_window_list.txt', 'r').read()
    windows = wmctrl._windows(window_list)
    window_ids = [window.window_id for window in windows]
    expected_window_ids = [
        '0x02a00001', '0x03200002', '0x0280000a', '0x03200007', '0x03200011',
        '0x03200016', '0x0320001b', '0x0320001c', '0x02c000b3', '0x03e0000c',
        '0x04000155', '0x042000b3', '0x04400051', '0x04c00001', '0x05000012',
        '0x03e005d4', '0x0580002b', '0x05600003']

    assert window_ids == expected_window_ids


def test_focus():
    """Test that Window._focus() calls _run() as we expect it to."""
    window = wmctrl.Window('test_window_id', 'desktop', 'pid',
                           'Navigator.Firefox', 'machine', 'title')
    mock_run_function = mock.MagicMock()

    window._focus(mock_run_function)

    assert mock_run_function.called_once_with('wmctrl -i -a test_window_id')


def test_focused_window():
    """Test that _focused_window() returns the right Window."""
    # Example output from wmctrl -a :ACTIVE: -v
    wmctrl_output = '''envir_utf8: 1
Using window: 0x03e00008
'''
    matching_window = wmctrl.Window('0x03e00008', 'desktop', 'pid', 'wm_class',
                                    'machine')
    windows = [
        wmctrl.Window('window_id', 'desktop', 'pid', 'wm_class', 'machine'),
        matching_window,
        wmctrl.Window('second_window_id', 'desktop', 'pid', 'wm_class',
                      'machine'),
    ]
    assert wmctrl._focused_window(wmctrl_output, windows) == matching_window


flitter/runraisenext.py
#!/usr/bin/env python3
"""A script for launching apps and switching windows."""
import sys
import argparse
import subprocess
import json
import os
import pickle
import re

from flitter import ewmh_window


def run(command):
    """Run the given shell command as a subprocess."""
    subprocess.call(command, shell=True)


def run_window_spec_command(window_spec, run_function):
    """Run the command from the given window spec, if it has one.

    :param window_spec: the window spec whose command to run
    :type window_spec: dict

    :param run_function: the function to call to run the command
    :type run_function: callable

    """
    command = window_spec.get('command')
    if command:
        run_function(command)


def focus_window(window):
    """Focus the given window.

    :param window: the window to focus
    :type window: wmctrl.Window

    """
    window.focus()


def get_all_window_specs_from_file(file_):
    """Return a list of all the window specs from the given config file."""
    specs = json.loads(
        open(os.path.abspath(os.path.expanduser(file_)), 'r').read())["specs"]
    lowercased_specs = {}
    for key in specs:
        assert key.lower() not in lowercased_specs
        lowercased_specs[key.lower()] = specs[key]
    return lowercased_specs


def get_window_spec_from_file(alias, file_):
    """Get the requested window spec from the config file.

    :rtype: dictionary

    """
    specs = get_all_window_specs_from_file(file_)
    spec = specs[alias.lower()]
    return spec


def get_ignore_from_file(file_):
    """Return the list of window specs to ignore from the given config file."""
    return json.loads(
        open(os.path.abspath(os.path.expanduser(file_)), 'r').read())[
            "ignore"]


def _load(path):
    """Helper function to load an object from persistent storage.

    This is a wrapper for pickle.load(), we wrap it to make it easy for tests
    to patch it and mock out the filesystem.

    """
    with open(path, "rb") as file_:
        return pickle.load(file_)


def _dump(obj, path):
    """Helper function to persist an object.

    This is a wrapper for pickle.dump(), we wrap it to make it easy for tests
    to patch it and mock out the filesystem.

    """
    with open(path, "wb") as file_:
        pickle.dump(obj, file_)


def pickle_path():
    """Return the path to the file we use to track windows in mru order."""
    return os.path.abspath(os.path.expanduser("~/.flitter.pickle"))


def sorted_most_recently_used(current_window_list):
    """Return the given list of open windows in most-recently-used order.

    :param current_window_list: the list of currently open windows,
        in any order
    :type current_window_list: list of Window objects

    :returns: the given list of currently opened windows, sorted into
        most-recently-used-first order
    :rtype: list of Window objects

    """
    try:
        pickled_window_ids = _load(pickle_path())
    except (IOError, EOFError):
        pickled_window_ids = []

    pickled_window_list = [ewmh_window.Window.window(window_id)
                           for window_id in pickled_window_ids]

    # Remove windows that have been closed since the last time we ran.
    pickled_window_list = [w for w in pickled_window_list if w is not None]

    # Add windows that have been opened since the last time we ran to the front
    # of the list.
    new_windows = []
    for window in current_window_list:
        if window not in pickled_window_list:
            new_windows.append(window)
    pickled_window_list = new_windows + pickled_window_list

    return pickled_window_list


def update_pickled_window_list(open_windows, newly_focused_window):
    """Move the newly focused window to the top of the cached list of windows.

    We keep a cached list of windows in most-recently-used order so that
    when switching to a new app we can switch to the app's most recently used
    windows first.

    Each time after focusing a window we call this function to update the
    cached list on disk for the next time we run.

    """
    assert newly_focused_window in open_windows
    open_windows.remove(newly_focused_window)
    assert newly_focused_window not in open_windows, (
        "There shouldn't be more than one instance of the same window in "
        "the list of open windows")
    open_windows.insert(0, newly_focused_window)
    _dump([w.window_id for w in open_windows], pickle_path())


def matches(window, window_spec):
    """Return True if the given window matches the given window spec.

    False otherwise.

    A window spec is a dict containing items that will be matched against
    the window object's attributes, for example:

        {'window_id': '0x02a00001',
         'desktop': '0',
         'pid': '4346',
         'wm_class': '.Firefox',
         'machine': 'mistakenot',
         'title': 'The Mock Class - Mock 1.0.1 documentation - Firefox'}

    A window object matches a spec if it has an attribute matching each of
    the items in the spec.

    A spec doesn't have to contain all of the attributes. For example
    {'wm_class': '.Firefox'} will match all windows with a wm_class
    attribute matching ".Firefox".

    Attribute matching is done by looking for substrings. For example a
    wm_class of ".Firefox" will match a window with a wm_class of
    "Navigator.Firefox".

    Window specs can also contain a "command" key (the command to be run to
    launch the app if it doesn't have any open windows) - this key will be
    ignored and the window will match the spec as long as all the other keys
    match.

    """
    for key in window_spec.keys():
        if key == 'command':
            continue
        if not re.match(window_spec[key], getattr(window, key, '')):
            return False
    return True


def matches_any(window, specs):
    """Return True if the given window matches any of the given specs."""
    for spec in specs:
        if matches(window, spec):
            return True
    return False


def _unvisited_windows(matching_windows, open_windows):
    """Return the list of matching windows that we haven't looped through yet.

    When we're looping through the windows of an app there's a continuous
    sequence of the app's windows at the top of the list of all open windows
    (which is sorted in most-recently-focused order). These are the windows
    from the app that we've already looped through.

    This function returns the app's windows that aren't part of this list:
    the ones we haven't looped through yet.

    May return an empty list.

    """
    visited_windows = []
    for window in open_windows:
        if window in matching_windows:
            visited_windows.append(window)
        else:
            break
    return [w for w in matching_windows if w not in visited_windows]


def _get_other_windows(windows, specs):
    """Return the windows that don't match any of the specs.

    Return the list of Window objects from windows that don't match any of the
    window spec dicts in specs. These are called the "other" windows.

    """
    return [w for w in windows if not matches_any(w, specs)]


def runraisenext(window_spec, run_function, open_windows, focused_window,
                 focus_window_function, others=False, window_specs=None,
                 ignore=None, current_desktop=False, ignore_minimized=False,
                 return_matching=False):
    """Either run the app, raise the app, or go to the app's next window.

    Depending on whether the app has any windows open and whether the app is
    currently focused.

    :param window_spec: the window spec to match against open windows
    :type window_spec: dict

    :param run_function: the function to use to run window spec commands
    :type run_function: callable taking one argument: the window spec

    :param open_windows: the list of open windows
    :type open_windows: list of Window objects

    :param focused_window: the currently focused window, should be one of the
        Window objects from open_windows
    :type focused_window: Window

    :param focus_window_function: the function to call to focus a window
    :type focus_window_function: callable taking one argument: a Window object
        representing the window to be focused

    :param others: If ``True`` go to the next "other" window ("other" windows
        are windows that don't match any of the given window specs) instead of
        looking for the next window that matches a particular window spec.
        The ``window_spec`` and ``run_function`` arguments aren't used if
        ``others`` is ``True``.

    :param window_specs: A list of all the window specs from the user's config
        file. This is only needed if ``others=True`` is given.
    :type window_specs: list of dicts

    :param ignore: A list of window specs matching windows that should be
        ignored and never focused. This is used to skip things like desktop
        windows, taskbars, etc.
    :type ignore: list of dicts

    :param current_desktop: Only raise windows from the current desktop
        (optional, default: False)
    :type current_desktop: bool

    :param ignore_minimized: Don't raise minimized windows
        (optional, default: False)
    :type ignore_minimized: bool

    :param return_matching: Just return the matching windows, don't run or
        raise anything (optional, default: False)
    :type return_matching: bool

    """
    def _focus_window(window):
        """Call focus_window_function() on the given window.

        Also moves the newly-focused window to the top of the
        most-recently-used-windows list, which we want to do whenever we
        focus a window.

        """
        focus_window_function(window)
        update_pickled_window_list(open_windows, window)

    if not ignore:
        ignore = []

    open_windows = sorted_most_recently_used(open_windows)

    # If no window spec options were given, just run the command
    # (if there is one).
    if (not others and
            'id' not in window_spec and
            'desktop' not in window_spec and
            'pid' not in window_spec and
            'wm_class' not in window_spec and
            'machine' not in window_spec and
            'title' not in window_spec):
        run_window_spec_command(window_spec, run_function)
        return

    # If there are no open windows, just run the command (if there is one).
    if not open_windows:
        run_window_spec_command(window_spec, run_function)
        return

    if others:
        matching_windows = _get_other_windows(open_windows, window_specs)
    else:
        matching_windows = [window for window in open_windows
                            if matches(window, window_spec)]

    matching_windows = [w for w in matching_windows
                        if not matches_any(w, ignore)]

    if current_desktop:
        current_desktop_ = ewmh_window.current_desktop()
        matching_windows = [w for w in matching_windows
                            if w.desktop == current_desktop_]

    if ignore_minimized:
        matching_windows = [w for w in matching_windows if not w.minimized]

    if return_matching:
        return matching_windows

    if not matching_windows:
        # The requested app is not open, launch it.
        run_window_spec_command(window_spec, run_function)
    elif focused_window not in matching_windows:
        # The requested app isn't focused. Focus its most recently used window.
        _focus_window(matching_windows[0])
    elif len(matching_windows) == 1 and focused_window in matching_windows:
        # The app has one window open and it's already focused, do nothing.
        pass
    else:
        # The app has more than one window open, and one of the app's windows
        # is focused. Loop to the app's next window.
        assert focused_window in matching_windows and len(matching_windows) > 1

        unvisited = _unvisited_windows(matching_windows, open_windows)
        if unvisited:
            _focus_window(unvisited[0])
        else:
            _focus_window(matching_windows[-1])


class ConfigFileError(Exception):
    pass


def _config_file_path(args):
    """Return the absolute path to the config file to use.

    This first uses the config file given with the -f/--f command-line
    argument (which defaults to ~/.flitter.json if not given).
    If that file doesn't exist it falls back on the default flitter.json file
    that ships with Flitter in the same directory as this Python module.
    Failing that it crashes.

    """
    path = os.path.abspath(os.path.expanduser(args.file))
    if os.path.isfile(path):
        return path

    default_path = os.path.join(
        os.path.split(sys.modules[__name__].__file__)[0], "flitter.json")
    if os.path.isfile(default_path):
        return default_path

    raise ConfigFileError(
        "Couldn't find either the config file {file} or the default config "
        "file {default_file}".format(file=path, default_file=default_path))


def parse_command_line_arguments(args):
    """Parse the command-line arguments."""
    parser = argparse.ArgumentParser(
        description="a script for launching apps and switching windows",
        add_help=True)

    window_spec_args = parser.add_argument_group("window spec")
    window_spec_args.add_argument(
        '-i', '--id', dest="window_id",
        help="the window ID to look for, e.g. 0x0180000b")
    window_spec_args.add_argument(
        '-d', '--desktop', help="the desktop to look for windows on, e.g. 1")
    window_spec_args.add_argument(
        '-p', '--pid', help="the pid to look for, e.g. 3384")
    window_spec_args.add_argument(
        '-w', '--wm_class',
        help="the WM_CLASS to look for, e.g. Navigator.Firefox")
    window_spec_args.add_argument(
        '-m', '--machine', help="the client machine name to look for")
    window_spec_args.add_argument(
        '-t', '--title',
        help='the window title to look for, e.g. "wmctrl - A command line '
        'tool to interact with an EWMH/NetWM compatible X Window Manager. - '
        'Mozilla Firefox"')

    parser.add_argument(
        '-c', '--command',
        help="the command to run to launch the app, if no matching windows "
        "are found, e.g. firefox")

    parser.add_argument(
        'alias', nargs='?',
        help="the alias of a window spec from the config file to use for "
        "matching windows")

    parser.add_argument(
        "-f", "--file", help="use a custom config file path",
        default="~/.flitter.json")

    parser.add_argument(
        "-o", "--others",
        help="raise \"other\" windows (windows that don't match any window "
             "spec in the config file), instead of windows matching a "
             "particular window spec",
        action="store_true")

    parser.add_argument(
        "--current-desktop",
        help="only raise windows on the currently active desktop",
        action="store_true")

    parser.add_argument(
        "--ignore-minimized",
        help="don't raise minimized windows",
        action="store_true")

    parser.add_argument(
        "--print-matching",
        help="just print the matching windows to standard out, don't run or "
             "raise anything",
        action="store_true")

    args = parser.parse_args(args)

    if args.window_id is not None:
        if (args.desktop or args.pid or args.wm_class or args.machine or
                args.title):
            parser.exit(status=1,
                        message="A window ID uniquely identifies a window, "
                        "it doesn't make sense to give the -i, --id argument "
                        "at the same time as any other window spec arguments")

    if args.others:
        if (args.command or args.window_id or args.desktop or args.pid or
                args.wm_class or args.machine or args.title):
            parser.exit(
                "The -o/--others argument can't be used at the same time as "
                "-c/--command, -i/--id, -d/--desktop, -p/--pid,-w/--wm_class, "
                "-m/--machine or -t/--title")

    try:
        config_file_path = _config_file_path(args)
    except ConfigFileError as err:
        parser.exit(err.message)

    # Form the window spec dict.
    if args.alias:
        window_spec = get_window_spec_from_file(args.alias, config_file_path)
    else:
        window_spec = {}
    if args.window_id is not None:
        window_spec['id'] = args.window_id
    if args.desktop is not None:
        window_spec['desktop'] = args.desktop
    if args.pid is not None:
        window_spec['pid'] = args.pid
    if args.wm_class is not None:
        window_spec['wm_class'] = args.wm_class
    if args.machine is not None:
        window_spec['machine'] = args.machine
    if args.title is not None:
        window_spec['title'] = args.title
    if args.command is not None:
        window_spec['command'] = args.command

    ignore = get_ignore_from_file(config_file_path)
    all_window_specs = get_all_window_specs_from_file(
        config_file_path).values()

    return (window_spec, all_window_specs, ignore, args.others,
            args.current_desktop, args.ignore_minimized, args.print_matching)


def main(args=None):
    if args is None:
        args = sys.argv[1:]
    window_spec, all_window_specs, ignore, others, current_desktop, ignore_minimized, print_matching = (
        parse_command_line_arguments(args))


    result = runraisenext(window_spec,
                          run,
                          ewmh_window.Window.windows(),
                          ewmh_window.Window.focused_window(),
                          focus_window,
                          others=others,
                          ignore=ignore,
                          window_specs=all_window_specs,
                          current_desktop=current_desktop,
                          ignore_minimized=ignore_minimized,
                          return_matching=print_matching)

    if print_matching:
        if result:
            return '\n'.join([str(w) for w in result])
    else:
        return result
